Task manager refactor revert pending wfs change

kdelee opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
We agreed last week that we want the TaskManager to deal with all pending tasks to keep all "soft blocking" logic contained to the DependencyGraph.

This does:
- [x] fix blocked_by logic for sliced jobs in depedency graph that was looking at wrong template ID
- [x] cherry-pick a fix @AlanCoding has for a fix to adding sliced jobs to the DepedencyGraph
- [x] move logic for moving pending workflows to running back to task manager
- [x] filter WorkflowApproval jobs out of the main get_tasks query so we are not grabbing them in DepdencyManager and TaskManager 
- [x] make get_tasks and related member functions set the lists of tasks as attributes on the classes and stop passing around the list of tasks
- [x] move calls to add newly running jobs to DepedencyGraph to start_task so we do it the same way for everything.

What I've tested (manually):
- [x] WorkflowJobs respect allow_simultaneous
- [x] SlicedJobs respect allow_simultaneous
- [x] regular jobs still start correctly
- [x] project update_on_launch still working

AlanCoding committed:
Fix unintended slice job blocking

kdelee committed:
Start pending workflows in TaskManager

we had tried doing this in the WorkflowManager, but we decided that
we want to handle ALL pending jobs and "soft blockers" to jobs with the
TaskManager/DependencyGraph and not duplicate that logic in the
WorkflowManager.

kdelee committed:
fix sliced jobs blocking logic in depedency graph

We have to look at the sliced job's unified_job_template_id
Now, task_blocked_by works for sliced jobs too.

AlanCoding commented on awx/main/scheduler/dependency_graph.py:
```diff
@@ -85,7 +96,11 @@ def job_blocked_by(self, job):
+                return self.get_item(self.WORKFLOW_JOB_TEMPLATES_JOBS, job.workflow_job_template_id)
+            elif job.unified_job_template_id:
+                # Sliced jobs are WorkflowJob type but do not have a workflow_job_template_id
+                return self.get_item(self.WORKFLOW_JOB_TEMPLATES_JOBS, job.unified_job_template_id)
```
Oh wow, good catch here. That makes the exclusivity rules complete, I just assumed it would be harder than this.

Let me just note - we don't have any testing for this, and it is a very notable absence.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -272,34 +276,13 @@ def timeout_approval_node(self):
-        running_workflow_tasks = self.get_tasks()
-        if len(running_workflow_tasks) > 0:
-            self.spawn_workflow_graph_jobs(running_workflow_tasks)
+        self.get_tasks()
```
I want to say this without taking a position on it:

It would be ideal that all subclasses of `TaskBase` have the same call pattern for `get_tasks()`. This may be the best, it may not be.

kdelee replied:
I think I addressed your concern, checkout latest commit

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -623,6 +615,13 @@ def process_pending_tasks(self, pending_tasks):
+            if isinstance(task, WorkflowJob):
+                # Previously we were tracking allow_simultaneous blocking both here and in DependencyGraph.
+                # Double check that using just the DependencyGraph works for Workflows and Sliced Jobs.
+                self.dependency_graph.add_job(task)
```
Not a critique of your PR, but we should consider moving the call to `self.dependency_graph.add_job` into the `start_task` method, as I cannot find a case where we call `start_task` without adding it to the dependency graph. I am aware that the `start_task` method does a lot of things, and it's extra mental energy to question whether this line in appropriate or not, when it should always be changing from pending to running.

kdelee replied:
agree

kdelee committed:
unify call pattern for get_tasks

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -572,10 +555,19 @@ def start_task(self, task, instance_group, dependent_tasks=None, instance=None):
+                execution_node_msg = f' and execution node {task.execution_node}' if task.execution_node else ''
+                logger.debug(
+                    f'Submitting job {task.log_format} controlled by {task.controller_node} to instance group {instance_group.name}{execution_node_msg}.'
+                )
```
Just marking for notes - this seems to bring us back to parity with `devel`.

The decisions here have significant system consequences and are non-obvious, so this is best. For example, from the cancel rework stuff, I know that workflows are assigned a `celery_task_id`. Because of that, I have to special-case workflows to not look for the process running under the dispatcher service. Nonetheless, that's accepted canonical behavior, so we shouldn't change it.

AlanCoding reviewed (approved):
This is great work. Down to the very small detail, I'm in strong agreement here.

kdelee committed:
add job to dependency graph in start task

We always add the job to the graph right before calling start task.
Reduce complexity of proper operation by just doing this in start_task,
because if you call start_task, you need to add it to the dependency
graph
